src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,462 @@
+                {
+                    return await _provider(cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    return ImmutableArray<TextSpan>.Empty;
+                }
+            }
+        }
+
+        private sealed class SolutionActiveStatementProviderCallback : IRemoteEditAndContinueService.ISolutionActiveStatementProviderCallback
+        {
+            private readonly SolutionActiveStatementSpanProvider _provider;
+
+            public SolutionActiveStatementProviderCallback(SolutionActiveStatementSpanProvider provider)
+                => _provider = provider;
+
+            /// <summary>
+            /// Remote API.
+            /// </summary>
+            public async Task<ImmutableArray<TextSpan>> GetSpansAsync(DocumentId documentId, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await _provider(documentId, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    return ImmutableArray<TextSpan>.Empty;
+                }
+            }
+        }
+
+        private IEditAndContinueWorkspaceService GetLocalService()
+            => _workspace.Services.GetRequiredService<IEditAndContinueWorkspaceService>();
+
+        public async Task StartDebuggingSessionAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                GetLocalService().StartDebuggingSession(solution);
+                return;
+            }
+
+            await client.RunRemoteAsync(
+                WellKnownServiceHubService.RemoteEditAndContinueService,
+                nameof(IRemoteEditAndContinueService.StartDebuggingSessionAsync),
+                solution,
+                Array.Empty<object>(),
+                callbackTarget: null,
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<IDisposable?> StartEditSessionAsync(
+            IDiagnosticAnalyzerService diagnosticService,
+            ActiveStatementProvider activeStatementProvider,
+            IDebuggeeModuleMetadataProvider debuggeeModuleMetadataProvider,
+            CancellationToken cancellationToken)
+        {
+            IDisposable result;
+            ImmutableArray<DocumentId> documentsToReanalyze;
+
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                GetLocalService().StartEditSession(activeStatementProvider, debuggeeModuleMetadataProvider, out documentsToReanalyze);
```

Comment:
this, in particular, seems weird to me.  i.e. say we start a debugging session remotely... then we call in later but arent' able to get a client... now we'd try to start an edit session locally?  would that jsut crash?

this is the first time i'm seeing a service with a lifetime associated with it, that makes multiple calls to get a client.  i don't feel like i have a good grasp of how that all works cleanly just yet.

Reply:
RemoteHostClient.TryGetClientAsync always succeeds or always returns null for given composition.